### ISSUE TITLE
Add option to QuantumCircuit.draw to omit final measurements

### ISSUE THREAD

**Issue by kevinsung** [Qiskit/qiskit#13581]
### What should we add?

Often the last operation in a circuit is a measurement of all the qubits, and this takes up a lot of space in the drawing. I would like to add an option to omit final measurements from the drawing.

**Comment by Cryoris** [comment]
The solution proposed in https://github.com/Qiskit/qiskit/issues/4149 which would just align all final measurements would also be nice, that way it is compact but we can still see the measurements.

**Comment by kevinsung** [comment]
> The solution proposed in [#4149](https://github.com/Qiskit/qiskit/issues/4149) which would just align all final measurements would also be nice, that way it is compact but we can still see the measurements.

Ah that's also a nice idea!

**Comment by enavarro51** [comment]
@Cryoris, @kevinsung, @jakelishman I can take a look at this, at least for the mpl drawer. It does bring up a number of design questions.

- In the following circuit, 
```
qr = QuantumRegister(2)
cr = ClassicalRegister(3)
qc = QuantumCircuit(qr, cr)
qc.x(0)
qc.h(0)
qc.measure(0, 0)
qc.x(1)
qc.h(1)
qc.measure(1, 1)
qc.x(1)
qc.measure(1, 2)
qc.h(1)
qc.draw()
```
![Image](https://github.com/user-attachments/assets/fae6e5f5-0c88-4f43-a5d1-05e270251664)

since the first 2 measures are at the same depth, it seems reasonable to make these without the double vertical line as pictured in #4149. Having said that, why not make the third measure the same and remove all the double lines to the classical bits since the bit will be indicated either in the measure box or under it?

- Taking this a step further,
```
qr = QuantumRegister(2)
cr = ClassicalRegister(3)
qc = QuantumCircuit(qr, cr)
with qc.if_test((cr, 2)):
    qc.x(0)
with qc.if_test(expr.equal(cr, 2)):
    qc.y(0)
qc.draw()
```
![Image](https://github.com/user-attachments/assets/dd79ef96-ee05-4531-b74e-51b284165754)

if we were to put `c6 == 2` under the `If` in the first red box, there would be no need for the double vertical lines here either. Which brings up the ultimate point, since the old `c_if` is going away and the measures can be done as above, why not do away with the entire classical bit section of the drawers? 

This would free up a lot of space both vertically and horizontially as we move to 100+ qubit drawings, and it should be workable for both the mpl and text drawers.

**Comment by Cryoris** [comment]
I like the idea for the measurements -- if we add the bit we measure into to the measurement box, there's no information in the double vertical lines and we can make the drawing a whole lot cleaner.

I'm less certain for the `c_if` bits, since the dots on the classical wire indicates the order in which operations are performed and thus do carry information.

**Comment by enavarro51** [comment]
I was planning on doing separate PRs for the measures and the ifs, so I'll move ahead on the measure PR if you can assign that to me.

In regards to the `c_if`, I was assuming the `If` change would be for 2.0, since c_if is being deprecated and removed in 2.0. Currently the `ControlFlowOp`s do not allow other operations to be displayed in the same layer (something that might be modified in the future if needed), so I think any order info is being preserved by the `IfElseOp`. Let me know if I'm missing something here.

**Comment by jakelishman** [comment]
Thanks Edwin!  (sorry for the slightly delayed response despite the tag - I just got back from paternity leave.)

I'm a bit nervous about committing to actually removing the clbits from the drawings (in Qiskit 2.0 land), but I do see the logic in it.  Would we have space to put the clbit information into/near the measurement box in the text drawer as well?

**Comment by enavarro51** [comment]
Yeah, I thought it might be a bit radical to remove the clbits, but realized if we don't need the double vertical lines, we don't really need the clbits.

For the mpl drawer, the clbit would appear in the measure box as a param would in a param gate. Might have to move the dial up slightly in the box, and the box would need to expand horizontally if the clbit name was long, but would probably be fine for the average use.

For the text drawer, the clbit would also have to appear where a param does now, which means it would add horizontal space to the drawing. For the original issue of final measurements of say 100 qubits, it clearly would use a whole lot less horizontal space, but individual measures would use more. My guess is on average, 2 to 3 old-style measures at the same depth would use about the same horizontal as 1 new style.

The same for the `If`s in the text drawer. It would add some horizontal to the `If` that's not using an expression already. This way `If`s with expressions and ones done with (bit, value) would appear the same.

Finally I just used 2.0 as a target because that's when `c_if` is supposed to go away. I was wondering if the look of a drawing changes but it still conveys the same information, is that a breaking change? Could it be done between major releases?

**Comment by enavarro51** [comment]
And one more final comment. The issue has so far been brought up for the mpl drawer. Do we have to do the mpl and text drawers the same? The mpl changes would be a clear improvement in space without loss of information. If that's not true for the text drawer, can we do the mpl and leave the text as is?

**Comment by enavarro51** [comment]
@jakelishman Here are a couple of samples of how the mpl changes would look.
![Image](https://github.com/user-attachments/assets/800170f4-4d5c-4761-b3b4-cf0a3c0d2c42)
![Image](https://github.com/user-attachments/assets/231c0126-cc1e-4a2a-b92e-fcf6aa714348)

**Comment by enavarro51** [comment]
Is there interest in pursuing this further?

**Comment by jakelishman** [comment]
Oh so sorry Ed - I never even saw these comments somehow.  If you're still up for it, I think those changes to the visualisers could be a very nice improvement, and fix the long-standing issue #4149.

**Comment by enavarro51** [comment]
Jake, I can move ahead with this. There are several design questions.

- Do we want to make the down line and arrow user configurable - perhaps with a `measure_clbit_arrow` parameter? If so, what would be the default?
- Do we want to remove the down line and arrow only for final measures?
- If so, how do we easily determine something is a final measure since the data sent to the drawers is in layer format?
- If we want to remove the line and arrow for all measures, I assume we would still need to make sure that gates or ControlFlowOps that are conditioned on the same clbit and follow the measure in QuantumCircuit.data order stay to the right of the measure?
- Do we want this for the text drawer as well?
- If so, If we remove the down line and arrow for the text drawer, it will expand how much horizontal room the measure will take, expanding the size of the drawing.

I think that's enough to get things moving.

**Comment by jakelishman** [comment]
Hey Ed -

I think an option is a good idea.  Perhaps `measure_arrows` - I think `clbit` might be redundant in there?  For this first implementation, let's default it to the current Qiskit behaviour, but let's make it configurable in the user config file so people can set it globally.  Then we can solicit a bit of feedback on it, and look to make it the default in the future, if it goes down well?

Personally I'd remove the arrow for all measures, if the option is set.  Alternatively, we _could_ have the `measure_arrows` take possible values like `False`, `True` and `"midcircuit"` but I think `True` and `False` would be fine to start.  It's trivial to see if something is a final measurement when everything's in a `DAGCircuit` form (which I believe the drawers do at _some_ point, right?).  I'm not very clear on the code structure of the visualisers - certainly it's possible to tell which measures are final from the `DAGCircuit` form, and that's the point at which we should tag and store the information.  I don't really know how the code looks, though.

We certainly need the layer structure to continue to represent the data-flow dependency, whether or not a vertical line is actually drawn, yeah.

I'd say it'd be ideal if it works with both the text and MPL drawers, but if you wanted to do only one at a time, that's no trouble - having it at all is better than neither.  I think for the text drawer, while it'll increase horizontal width of an individual box, it should let all the boxes stack on top of each other, so should end up thinner?  If it makes the drawing wider, then it likely isn't worth it.

**Comment by enavarro51** [comment]
Thanks, Jake. So I will proceed as follows,
- Make measure display change optional using True/False user param `measure_arrows`, defaulting to True.
- Add `user_config` option for user to set the default.
- If False, the change will be made to all measures.
- The same changes will be made to both the `mpl` and `text` drawers. The latex drawer will function as is.
